### PR TITLE
Make the /runs page fully React.js-y + Add autocomplete

### LINKED ISSUE #4: autocomplete dataset name
To ensure that entries that are made on the same dataset match in dataset name can we have this autocomplete or as a dropdown as already submitted options + new?

### PR DESCRIPTION
**Pull request by ihodes**
This concludes the rewrite of /runs using React.js, and adds typeahead to form fields on that page.

fixes #4 

[Review on Reviewable](https://reviewable.io/reviews/hammerlab/cycledash/463)

### CODE REVIEW

**Comment by danvk** [comment]
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/cycledash/463)



---

Cool! LMK when it's ready for review (and don't forget to update the screenshots!).

---

File: cycledash/static/css/runs.css [hunk 1]
```diff
@@ -5,7 +5,6 @@ main {
 }
 form#submit {
   margin: 0 auto 27px auto;
-  display: none;
   max-width: 500px;
 }
 form#submit .form-control {
```
Comment (ihodes) [inline cycledash/static/css/runs.css]: test

**Comment by ihodes** [comment]
@danvk Ready for your eyes now!

**Comment by danvk** [comment]
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/cycledash/463)

Reviewed files:
- cycledash/static/css/runs.css @ r2
- cycledash/static/css/style.css @ r1
- cycledash/static/js/CompletionUtils.js @ r2
- cycledash/static/js/runs.js @ r1
- cycledash/static/js/runs/components/LatestComments.js @ r1
- cycledash/static/js/runs/components/RunsPage.js @ r2
- cycledash/static/js/runs/components/SubmitRunForm.js @ r2
- cycledash/static/js/runs/runs.js @ r2
- cycledash/templates/macros/run_form.html @ r1
- cycledash/templates/runs.html @ r2
- cycledash/views.py @ r2
- gulpfile.js @ r1
- package.json @ r1
- tests/pdifftests/runs-info.png @ r2
- tests/pdifftests/runs.png @ r2
- workers/genotype_extractor.py @ r1

---

Looks good, a bunch of (mostly small) comments. I like the pervasive typeahead!

---

<sup>**[cycledash/static/css/runs.css, line 25 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1pKr1nayrQLJm36r_)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/css/runs.css#L25)):</sup>
should be the same as `margin: 29px 0;`

---

<sup>**[cycledash/static/js/runs/components/LatestComments.js, line 37 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1bfCsaXtyg68O6iLd)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/LatestComments.js#L37)):</sup>
any reason for the spaces around `{ ... }` here (there are no spaces in the JSX in `LatestComments`).

---

<sup>**[cycledash/static/js/runs/components/RunsPage.js, line 13 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1h36AV1oA-jCbJ4Jj)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/RunsPage.js#L13)):</sup>
Document what this is & its expected shape.

---

<sup>**[cycledash/static/js/runs/components/RunsPage.js, line 15 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1gyAuzxiNK7VJBqZ9)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/RunsPage.js#L15)):</sup>
Say something about what these are — I'm not sure what a "completion" is.

---

<sup>**[cycledash/static/js/runs/components/RunsPage.js, line 18 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1bz3knrXUNBTanIKR)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/RunsPage.js#L18)):</sup>
stick each of these on its own line

---

<sup>**[cycledash/static/js/runs/components/RunsPage.js, line 39 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1cH5tf20keEeHymua)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/RunsPage.js#L39)):</sup>
Split this into two functions (like `handleDrag{Over,Leave}`) or rename it -- it's fundamentally different from the other methods that start with `handle` (it returns a handler), so it should be named differently, too.

---

<sup>**[cycledash/static/js/runs/components/RunsPage.js, line 55 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1eKmOKzRrGbVot0YF)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/RunsPage.js#L55)):</sup>
`projectOptions`?

---

<sup>**[cycledash/static/js/runs/components/RunsPage.js, line 55 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1eTcmF8HsOYdF-_k5)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/RunsPage.js#L55)):</sup>
`[NO_FILTER].concat(projectNames).map` and kill the next line?

---

<sup>**[cycledash/static/js/runs/components/RunsPage.js, line 60 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1gdDkidrrjjV1A0HL)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/RunsPage.js#L60)):</sup>
`(run)` → `run`

---

<sup>**[cycledash/static/js/runs/components/RunsPage.js, line 65 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1gHzKZf2xEBQMFI0S)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/RunsPage.js#L65)):</sup>
Would it make sense to move the `RunDescription` into `Run` and have a `selected` property?

---

<sup>**[cycledash/static/js/runs/components/RunsPage.js, line 68 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1faf1Tf1l6wzEhHia)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/RunsPage.js#L68)):</sup>
`run.id` should be fine -- these keys don't need to be globally unique

---

<sup>**[cycledash/static/js/runs/components/RunsPage.js, line 78 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1hmzirm8Z_k9-B3cD)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/RunsPage.js#L78)):</sup>
just pass `completions` to `SubmitRunForm`?

---

<sup>**[cycledash/static/js/runs/components/RunsPage.js, line 108 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1gWK_EekL9xGivmeR)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/RunsPage.js#L108)):</sup>
GTK: React is cool with lists-of-lists

---

<sup>**[cycledash/static/js/runs/components/RunsPage.js, line 116 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1iOHpygv0wXF-VFyA)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/RunsPage.js#L116)):</sup>
`RunRow`? Makes it clear that this returns a `<tr>`. (similar thoughts for other components below)

---

<sup>**[cycledash/static/js/runs/components/RunsPage.js, line 146 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1ijRGPUS98aq_Tifq)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/RunsPage.js#L146)):</sup>
can't you just do this with `map`?

---

<sup>**[cycledash/static/js/runs/components/RunsPage.js, line 172 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1j56-a5i_Ni0a5bEU)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/RunsPage.js#L172)):</sup>
I was curious whether this would feel better if you pulled the logic out into an array. Take it or leave it, I don't prefer either strongly.

```javascript
var labelTypes = [
    ['validation_vcf', 'validation', 'Is a validation VCF'],
    ['tumor_bam_uri', 'tumor', 'Has an associated tumor BAM'],
    ['normal_bam_uri', 'normal', 'Has an associated normal BAM']
];
var labels = labelTypes.map((key, text, title) => {
  if (!run[key]) return [];
  return [<span className='label label-info'
          title={title}
          key={key}>
      {text}
  </span>];
});
```

---

<sup>**[cycledash/static/js/runs/components/RunsPage.js, line 209 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1kUBqsKAnhP7P30qd)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/RunsPage.js#L209)):</sup>
move `no-comment` class up into the `td` to avoid duplication?

---

<sup>**[cycledash/static/js/runs/components/SubmitRunForm.js, line 10 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1n1qgpTZmfjqiqhHb)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/SubmitRunForm.js#L10)):</sup>
still not released :(

---

<sup>**[cycledash/static/js/runs/components/SubmitRunForm.js, line 36 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1nLeogIdYBuvH7r6J)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/SubmitRunForm.js#L36)):</sup>
indent

---

<sup>**[cycledash/static/js/runs/components/SubmitRunForm.js, line 74 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1oNJ3GGwoLIOFdOqT)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/SubmitRunForm.js#L74)):</sup>
`CompletableTextInput`? There's a lot more going on in this component than the name would imply (typeahead, file upload).

---

<sup>**[cycledash/views.py, line 69 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1odmK_d6fS7ibbTWD)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/views.py#L69)):</sup>
revert

---

<sup>**[cycledash/views.py, line 84 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1oieLPuofn3Fj_hqe)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/views.py#L84)):</sup>
the `_` seems excessive for a local function definition

---

<sup>**[cycledash/views.py, line 84 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1otXNRVExgToexLmn)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/views.py#L84)):</sup>
I see where the `pluck` is, but I don't see the `unique`. Do you want to change `list` to `set` on the next line?

---

<sup>**[cycledash/views.py, line 179 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1p-iPzWqnL5WuKiAV)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/views.py#L179)):</sup>
rv

---

**Comment by ihodes** [comment]
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/cycledash/463)







---

<sup>**[cycledash/static/js/runs/components/LatestComments.js, line 37 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1bfCsaXtyg68O6iLd)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/LatestComments.js#L37)):</sup>
Only a poorly executed regexp search & replace! Fixed

---

<sup>**[cycledash/static/js/runs/components/RunsPage.js, line 66 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1gHzKZf2xEBQMFI0S)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/RunsPage.js#L66)):</sup>
Unfortunately this doesn't really work + blah (cf. discussion with Tavi when I wrote this code: https://reviewable.io/reviews/hammerlab/cycledash/450#-JiI1TtIXESNm_qkCp6B

---

<sup>**[cycledash/static/js/runs/components/RunsPage.js, line 69 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1faf1Tf1l6wzEhHia)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/RunsPage.js#L69)):</sup>
It needs a prefix so that it doesn't collide with a Run of the same ID (of which there will always be one in this array).

---

<sup>**[cycledash/static/js/runs/components/RunsPage.js, line 173 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1j56-a5i_Ni0a5bEU)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/RunsPage.js#L173)):</sup>
I like your idea better; using this code. (FWIW, you can't do restructuring with fat-arrow functions; needs to be `function([key, text, title])`)

---

<sup>**[cycledash/static/js/runs/components/SubmitRunForm.js, line 74 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1oNJ3GGwoLIOFdOqT)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/static/js/runs/components/SubmitRunForm.js#L74)):</sup>
There is more going on, but I think that's made clear to the reader in the optional props being passed; it's the `TextInput` we deserve! 

---

<sup>**[cycledash/views.py, line 84 \[r2\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj1otXNRVExgToexLmn)** ([raw file](https://github.com/hammerlab/cycledash/blob/879f8acd317fc382ae9a5af07e6779d7174ba91c/cycledash/views.py#L84)):</sup>
Aha! I am using a set comprehension, not a list comprehension, so it'll be unique because of that.

---

**Comment by danvk** [comment]
Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/cycledash/463)



---

Looks good!

---

<sup>**[cycledash/static/js/runs/components/RunsPage.js, line 128 \[r4\]](https://reviewable.io:443/reviews/hammerlab/cycledash/463#-Jj28MDswDlVLvDfzmua)** ([raw file](https://github.com/hammerlab/cycledash/blob/e1d4ac003765f54d604aaab2584595ddf0c79d10/cycledash/static/js/runs/components/RunsPage.js#L128)):</sup>
same comment about `{ spaces }`

---

**Pull request merged**